### ISSUE TITLE
Weapon projectile is offset at high speed

### ISSUE THREAD

**Issue by tibetiroka** [endless-sky/endless-sky#5857]
**Describe the bug**
The image of the weapons' projectile is offset from the ship's position. This offset is based on the ship's speed, so it is not normally noticable. However, if your ship reaches _considerable_ speed, the offset will be clearly visible. In the example I used Dragonflame Cannons because 1. they have a "beam" instead of a normal projectile which makes it easier to notice, and 2. they help you gather speed. If you shoot with it 30-40-50 times rapidly (like, once per second) you should reach this speed.

Firing while going backwards:
![Firing while going backwards](https://user-images.githubusercontent.com/68112292/115149045-8398cb00-a062-11eb-9b70-bea67eb4ca79.png)

Firing while going forward:
![Firing while going forward](https://user-images.githubusercontent.com/68112292/115149056-94494100-a062-11eb-9774-206932cd4dbc.png)

Firing while stationary:
![Firing while stationary](https://user-images.githubusercontent.com/68112292/115149100-d07ca180-a062-11eb-9c6d-469202c123b5.png)

The offset of the weapon mount and the projectile seems to be a constant if you are not accelerating, but it is difficult to test with this setup.

**To Reproduce**
1. Accelerate your ship to (extremely) high speed. 
2. Zoom in fully (optional but really helps)
3. Fire your gun.
4. See the bug

**Expected behavior**
The projectile should either render at the same place every time, or it should 'travel away' from the ship. But rendering with an offset and keeping that relative location doesn't seem to be intended.

**Screenshots**
![Additional screenshot](https://user-images.githubusercontent.com/68112292/115149537-e3907100-a064-11eb-8cb6-36d5ddb2bafb.png)
I uploaded this additional screenshot because it was really interesting: the dragonflame cannon's projectile was offset, but the skylance's wasn't. Maybe this is because the skylance has continuous fire?

**System (please complete the following information):**
 - OS: Windows 10 Pro, version 20H2 build 19042.928
 - Game Source: Steam
 - Version: 0.9.13
 
**Link to save file**
I don't think it is necessary, but I'll post the config for my test ship: [pastebin link](https://pastebin.com/k95bctZ8)

**Additional context**
In the example I was using the high dpi plugin, but I could reproduce the issue without it.

**Comment by vitalchip** [comment]
The Dragonfalme Cannon doesn't have a projectile sprite. What you see on screen are effects. This will be why, as you noted, the Skylance beam (being a projectile sprite) looks fine.

The dragonflame effect persists for 21 frames and has velocity scale 1 so the most likely situation is that your ship has a small acceleration at the time of firing and either moves away from or moves toward the effect.

**Comment by tibetiroka** [comment]
In my testing the acceleration was pretty much the same every time (it was based on the weapon's force), but the screen effect was more offset at high speed - so it seems unlikely that it is rendered incorrectly because the ship accelerated. Unless there is a bug that causes the ship to have higher acceleration for the same force at a higher speed, I don't see how this could happen. 
 I'd say it is much more likely that the position of the effect was calculated before the ship moved, because then the offset would be based on the current velocity.

**Comment by vitalchip** [comment]
I think this is  just going to be order of operations. The effect is placed and then the ship is moved. At high speeds it moves further so you see a larger gap. Ordinarily the amount a ship can be expected to move in a frame would be a few pixels. A combination of a very high speed and a persisting effect pretending to be a beam weapon will show a larger gap.

**Comment by Amazinite** [comment]
This seems related to #4260.

**Comment by tibetiroka** [comment]
@vitalchip yep, this is what I meant.

@Amazinite I agree, it is probably the same thing. Should have been fixed, but I guess nothing is perfect.

**Comment by vitalchip** [comment]
This commit fixed the afterburner issue. https://github.com/endless-sky/endless-sky/commit/d4e4bb26b1d022bfa44d5392c87b37e387d18e54

**Comment by vitalchip** [comment]
Could this be happening because new visuals are appended to the object list after visuals are moved for this step. So going through Engine::CalculateStep() 

1. Commands actioned (so weapons fired and fire effect created by Hardpoint)
2. Ship moved
3. Visuals moved
4. New visuals appended. (So the visual created on step 1 is now appended to be drawn but the position is where the ship *was* before step 2. )

I think, some way or other the visual must be getting drawn at a 'previous' ship position since it leads or lags depending on the ship direction wrt to the line of fire.

**Comment by vitalchip** [comment]
Right, appending the new visuals before moving them just shifted the problem forwards or backwards. What did solve it was changing the start point of the fire effect to *not* include the -.5 *ship.Velocity() that the projectiles get. I haven't checked for any negative effects on anything else yet but it does fix this particular issue.
<img width="741" alt="Screenshot 2021-04-18 at 21 44 18" src="https://user-images.githubusercontent.com/65342577/115160272-5282bf80-a08f-11eb-99ba-39c151317279.png">

**Comment by vitalchip** [comment]
Before and after clips. Moving at speed of 5000 (top right, under flagship HUD) and turned at ~90 degrees to direction of motion for firing.

https://user-images.githubusercontent.com/65342577/115160333-9bd30f00-a08f-11eb-87d3-19cce8ebbb9c.mov


https://user-images.githubusercontent.com/65342577/115160342-a2fa1d00-a08f-11eb-8d29-4279783c6214.mov

**Comment by oo13** [comment]
Probably #5479 is related to this issue.

**Comment by ravenshining** [comment]
I'm not sure this is solely an appearance thing, I once abused this bug with explosives in asteroid fields to accelerate my ship in odd ways - that is, I would fire a weapon to be moved by the explosion, and not only would the projectile appear in the wrong place, but would apply hit force from there as well.
